80-tgyuuAn

tgyuuAn opened the pull request:
<!-- PR은 최대한 다른 사람이 알아보기 쉽도록 자세히 써주세요. 특히 수도 코드 부분은 더더욱요...!!-->

## 🔗 문제 링크
<!-- 해결한 문제의 링크를 올려주세요. -->
[최소비용 구하기 2](https://www.acmicpc.net/problem/11779)

## ✔️ 소요된 시간
<!-- 문제를 해결하는데 소요된 시간을 적어주세요. -->
30분

## ✨ 수도 코드
<!-- 내가 작성한 코드를 모르는 사람이 봐도 이해할 수 있도록 글로 쉽게 풀어서 설명해주세요. -->

![image](https://github.com/user-attachments/assets/023165a2-04db-452f-9792-10de5da6ff74)

알고리즘,,, 거의 한달만에 푸는 것 같네용 _(경진대회 제외하구!)_

롱타임노씨입니다!

소소하게 골드 2~3 중에서 모색하다가 문제 이름 보고 왠지 DP, 그리디 일 것 같아서 골랐는데 다른 알고리즘이었네요 ,,

<br><br><br>

문제는 되게 간단합니다.

![image](https://github.com/user-attachments/assets/f1a88a15-3195-4e82-9702-3711b83f10b3)

위 지문을 읽자마자, **A에서 B도시로 가는 버스 최소 비용을 구해라** 라는 말을 보자마자 다익스트라 라는 것을 직감하였습니다.

심지어 노드의 개수도 **10만**이고 간선의 개수가 **1000** 이기 때문에, $10만 * log(1000) = 100만$ 이므로 시간복잡도도 세이프이죠!

맞아요 문제 그냥 다익스트라 풀면 풀립니다.

하지만 문제는 여기서 한 가지 트릭을 더 걸었어요.

<br><br><br>

![image](https://github.com/user-attachments/assets/694e1664-91c8-4c4a-8670-67930c0f7e8d)

출력에서 **최소 비용을 가지는 경로에서 도시의 개수 및 도시의 경로**를 출력해야해요.

<br><br><br>

이걸 처음 봤으면 난해했을 수도 있습니다,,

_`각 노드마다 히스토리를 모두 저장해야하나...?`_ 하고 말이에요.

하지만 저는 https://github.com/AlgoLeadMe/AlgoLeadMe-1/pull/85 를 바로 떠올릴 수 있었고,

**해당 노드로 갈 때 이전 노드만 기록**해준 뒤, 마지막에 도착지점으로부터 시작 지점까지 타고타고 올라가면 완성입니다!

<br><br><br>

이 문제,, 처음 접하면 조금 당황스러울 수도 있을 것 같아요.

홍주님 몇분 걸렸는지 말해주세용~!!!!



## 📚 새롭게 알게된 내용
<!-- 새롭게 알게된 내용이 있다면 작성 해주시고 출처를 남겨주세요. -->

H0ngJu reviewed (approved):
1. list 출력 포맷적용 안해서
2. 다익스트라에서 초기화가 잘못돼서

오늘도 질게 테케를 통해 풀었습니다 하핫
1시간 10분 쫌 넘게 풀었습니둥 ...!>!!>!>!>!>!>
PR에서 말씀하신대로 경로출력하는게 낯설긴 했는데 여차저차해서 해결했습니다!!

#85 보니까 이전경로만을 저장해서 역추적.. 하는 방법이 있었군요 ....
저는 냅다 배열에 다 저장하고, 새 노드 경로를 추가하여 갱신하는 식으로 진행했습니다
힙을 사용했으면 경로 추적도 그렇고 다익스트라도 좀더 간단하게 풀 수 있었겠네용 ..

```python
# 9시 20분 시작

import sys

def input() : return sys.stdin.readline().rstrip()

n = int(input()) # 1000
m = int(input()) # 100000
info = [ [] for _ in range(n+1)]
# 도시번호(출발지) 도시번호(도착지) 버스비용
for i in range(m):
    s, e, c = map(int, input().split())
    info[s].append((e,c))
start_city, end_city = map(int, input().split())
cost = [1e9 for _ in range(n+1)] 
path = [[] for _ in range(n+1)]
visited = [False for _ in range(n+1)]

def get_min_cost_node():
    min_cost = 1e9
    idx = 0
    for i in range(1,n+1):
        if cost[i] < min_cost and not visited[i]:
            min_cost = cost[i]
            idx = i
    return idx

# 다익스트라임
def Dijikstra(start):
    cost[start] = 0
    visited[start] = True
    for e, c in info[start]:
        cost[e] = min(cost[e], c) #여러 경로가 있을 경우 최솟값으로 초기화
    for i in range(n-1):
        node = get_min_cost_node()
        visited[node] = True
        for e, c in info[node]:
            if cost[e] > cost[node] + c:
                cost[e] = cost[node] + c  
                path[e] = path[node] + [node]

if start_city == end_city:
    print(0)
    print(1)
    print(start_city)
else: 
    Dijikstra(start_city)
    print(cost[end_city])
    path[end_city] = [start_city] + path[end_city] + [end_city]
    print(len(path[end_city]))  
    print(" ".join(map(str, path[end_city])))
```

tgyuuAn commented:
> 1. list 출력 포맷적용 안해서
> 2. 다익스트라에서 초기화가 잘못돼서
> 
> 오늘도 질게 테케를 통해 풀었습니다 하핫 1시간 10분 쫌 넘게 풀었습니둥 ...!>!!>!>!>!>!> PR에서 말씀하신대로 경로출력하는게 낯설긴 했는데 여차저차해서 해결했습니다!!
> 
> #85 보니까 이전경로만을 저장해서 역추적.. 하는 방법이 있었군요 .... 저는 냅다 배열에 다 저장하고, 새 노드 경로를 추가하여 갱신하는 식으로 진행했습니다 힙을 사용했으면 경로 추적도 그렇고 다익스트라도 좀더 간단하게 풀 수 있었겠네용 ..
> 
> ```python
> # 9시 20분 시작
> 
> import sys
> 
> def input() : return sys.stdin.readline().rstrip()
> 
> n = int(input()) # 1000
> m = int(input()) # 100000
> info = [ [] for _ in range(n+1)]
> # 도시번호(출발지) 도시번호(도착지) 버스비용
> for i in range(m):
>     s, e, c = map(int, input().split())
>     info[s].append((e,c))
> start_city, end_city = map(int, input().split())
> cost = [1e9 for _ in range(n+1)] 
> path = [[] for _ in range(n+1)]
> visited = [False for _ in range(n+1)]
> 
> def get_min_cost_node():
>     min_cost = 1e9
>     idx = 0
>     for i in range(1,n+1):
>         if cost[i] < min_cost and not visited[i]:
>             min_cost = cost[i]
>             idx = i
>     return idx
> 
> # 다익스트라임
> def Dijikstra(start):
>     cost[start] = 0
>     visited[start] = True
>     for e, c in info[start]:
>         cost[e] = min(cost[e], c) #여러 경로가 있을 경우 최솟값으로 초기화
>     for i in range(n-1):
>         node = get_min_cost_node()
>         visited[node] = True
>         for e, c in info[node]:
>             if cost[e] > cost[node] + c:
>                 cost[e] = cost[node] + c  
>                 path[e] = path[node] + [node]
> 
> if start_city == end_city:
>     print(0)
>     print(1)
>     print(start_city)
> else: 
>     Dijikstra(start_city)
>     print(cost[end_city])
>     path[end_city] = [start_city] + path[end_city] + [end_city]
>     print(len(path[end_city]))  
>     print(" ".join(map(str, path[end_city])))
> ```

다익스트라임 무엇 ㅋㅋㅋㅋㅋㅋㅋㅋㅋ
